stl_text/datamodule/translation.py

```diff
@@ -0,0 +1,167 @@
+        batch_size_sequences: Optional[int] = 16,
+        batch_size_tokens: Optional[int] = None,
+        source_text_transform: Optional[nn.Module] = None,
+        target_text_transform: Optional[nn.Module] = None,
+        num_dataloader_workers: int = 1,
+        num_setup_workers: int = 4,
+    ):
+        super().__init__()
+        self.source_data = source_data
+        self.target_data = target_data
+        self.source_vocab = source_vocab
+        self.target_vocab = target_vocab
+        self.batch_size_sequences = batch_size_sequences
+        self.batch_size_tokens = batch_size_tokens
+        self.source_text_transform = source_text_transform
+        self.target_text_transform = target_text_transform
+        self.num_dataloader_workers = num_dataloader_workers
+        self.num_setup_workers = num_setup_workers
+
+        # TODO should torchtext vocab define pad, bos, eos, etc?
+        # TODO otherwise take these in as input arguments?
+        self.source_pad_idx = 0
+        self.target_bos_idx = 0
+        self.target_pad_idx = -100
+
+        assert self.source_pad_idx >= 0, \
+            f"source_pad_idx must be >= 0 for nn.Embedding ({self.source_pad_idx})"
+        assert self.target_bos_idx >= 0, \
+            f"target_bos_idx must be >= 0 for nn.Embedding ({self.target_bos_idx})"
+
+        if batch_size_tokens is not None:
+            raise NotImplementedError
+
+        if batch_size_sequences is None and batch_size_tokens is None:
+            raise ValueError
+
+        self.datasets = {}
+
+    def setup(self, stage):
+        for split in ("train", "valid", "test"):
+            src = self.source_data[split].map(
+                function=lambda x: {"token_ids": self.source_text_transform(x)},
+                input_columns="text",
+                num_proc=self.num_setup_workers,
+            )
+            src.set_format(type="torch", columns=["token_ids"])
+            tgt = self.target_data[split].map(
+                function=lambda x: {"token_ids": self.target_text_transform(x)},
+                input_columns="text",
+                num_proc=self.num_setup_workers,
+            )
+            tgt.set_format(type="torch", columns=["token_ids"])
+
+            self.datasets[split] = DictDataset({
+                "source": src,
+                "target": tgt,
+            })
+
+    def forward(self, source_text_batch: List[str], target_prefix_batch: Optional[List[str]] = None) -> torch.Tensor:
+        if target_prefix_batch is not None:
+            raise NotImplementedError
+        data = [
+            {"source": {"token_ids": self.source_text_transform(text)}}
+            for text in source_text_batch
+        ]
+        return torch.utils.data.DataLoader(
+            data,
+            shuffle=False,
+            # TODO allow batch size to be configured here? do we even want batching?
+            batch_size=self.batch_size_sequences,
+            collate_fn=self.collate,
+        )
```

Comment:
What would be length for `source_text_batch`? I think even for batch inference, the length should be less than 1 thousand. So `DataLoader` is not needed here. A `for loop` might be enough